### DISCUSSION TITLE
Splitting the course

### DISCUSSION THREAD

**Discussion by dalonsoa** [ImperialCollegeLondon/grad_school_git_course#29] [Ideas]
One of the main conclusions of the 2022 evaluation is that the course, as it is, needs to be split in two - one version simpler for novice people and another one more advanced (intermediate level) for those who already has some experience.

This splitting will be effective from the Spring term 2023, but we need to define the content well in advance. Assuming each course will be 3h long in a single session - with a 10 min break in the middle - suggestion of the splitting will be:

**Git/GitHub Basics:**
- Nearly identical to the current course up to (and including) [working with remotes](https://imperialcollegelondon.github.io/grad_school_git_course/l2-02-remote_repositories/index.html).

**Git/GitHub Intermediate:**
- Current [Collaborating](https://imperialcollegelondon.github.io/grad_school_git_course/l2-03-collaborating/index.html) section
- Some advanced `git` commands: `reset`, `rebase`, working with multiple remotes, tagging versions
- Some advance GitHub functionality: Creating releases, continuous integration with GitHub actions (the basics)

Any comments, suggestions or other approaches on how to do this?

**Comment by dandavies99** [comment]
@ImperialCollegeLondon/rse-team 

As mentioned above and discussed in the meeting, it might be a good idea to split out the concepts of branches and working collaboratively into the intermediate course.

Proposed changes to the current course to form the basic course: 

- Setup
  - Leave as is
- Why use version control
  - Leave as is
- Committing and history
  - Leave as is
- ~Branching and merging~
  - Move to intermediate  
- Sharing your code
  - This section introduces GitHub, licenses, private/public repos etc. Leave as is. 
- Remote repositories
  - Simplify by reframing as single-branch workflows. i.e. remove references to "feature branch". There isn't too much of this anyway, but I think acknowledging there are such a thing is branches, but we're sticking with just one "main" branch for the beginners course should be ok. 

**Questions**
- How does the above sound for the basics course? 
- Are there any key concepts missing? 
- Are there any key concepts left in there that maybe should be moved to intermediate?

**Reply by AdrianDAlessandro** [reply]
I think Diego's question about a graphical tool is worth discussing. I think using a graphical tool is what will be most useful to people that only want the beginners course / said the existing one was too complicated. Do we maybe only use a graphical tool in this course and introduce the command line in the intermediate course?

**Reply by dandavies99** [reply]
**Trying to consolidate graphical tools discussion here:**

I certainly think we need to touch on GUIs because whenever I've seen a straw poll carried out amongst git users with options (i) I only use git at the command line (ii) I only use a GUI and (iii) I use a mixture, it's always overwhelmingly a (iii) majority with some (ii) and (i). So, realistically I don't think skipping GUIs is very representative of how most people interact with git.  

If we did everything from a graphical tool for the basics course, I worry it would be too challenging for learners to then do the intermediate course (a natural progression) having never covered git at the command line. Also the GUIs hide a lot of what is actually going on, e.g. having a "sync" button which does some combo of push and pull, so some of the fundamentals are more easily covered from the command line. 

I suggest we keep the bulk of the course based at the command line, but allow some time at the end to cover:
- Links to a couple of GUI options, perhaps one IDE plugin (e.g. VS Code) and one or two others (e.g. GitHub Desktop, GitKraken).
- A short demo of one of these, showing how a few simple commands that we've already covered in the course are carried out (e.g. add, commit).
- A short demo of how the GUI can be linked to a remote repository.

What do you think @AdrianDAlessandro @dalonsoa @alexdewar?

**Reply by AdrianDAlessandro** [reply]
This sounds sensible to me!

**Reply by alexdewar** [reply]
Agreed. You could also use screenshots as illustrations for things like merging etc. as the output is often quite pretty with a GUI.

There's also a nice curated list of git GUIs on the git website: https://git-scm.com/downloads/guis/

**Reply by cc-a** [reply]
Based on the current timings for those sections the listed content for this course runs to about 2 hrs. So there is certainly time to look at some GUI's and probably some other significant item.

**Comment by dalonsoa** [comment]
I think this looks ok, as long as the concept of "branches" is indeed mentioned. I think it is important for them to finish the course knowing that they exist and what they might be useful for, even if that is not put into practice.

Are we going to mention a graphical tool like GitHub Desktop?

**Reply by dandavies99** [reply]
Agree re: branches - it would likely cause more confusion to try and avoid the topic, they're pretty fundamental. I think it'll save us a lot of time to just explain what they are, say that we're sticking with one "main" branch for this course, their practical uses are covered in depth in the intermediate course, and move on.

**Comment by alexdewar** [comment]
Not sure if I've raved to you about GitKraken before, but it's an excellent GUI tool and (I think) v user friendly! https://www.gitkraken.com/

You can get the Pro version if you're an academic.

**Comment by dandavies99** [comment]
# Ideas for intermediate course content

Let's get all possibilities down before prioritising and structuring. **Please add additional points below - you should be able to edit this comment!** ⬇️ 

- Branches
  - Concept of an experimental "feature" branch that can be abandoned or merged in (could take more or less as-is from existing course)
- Current [Collaborating](https://imperialcollegelondon.github.io/grad_school_git_course/l2-03-collaborating/index.html) section.
  - Any changes to this section...? 
- Some advanced git commands 
  - reset
  - stash
  - rebase, including merge vs rebase
  - working with multiple remotes _<-- is this a common situation?_
  - tagging versions
- Some advance GitHub functionality 
  - Creating releases 
  - Continuous integration with GitHub actions (the basics, e.g. running unit tests)
- Common git workflows
  - Other than "Fork and pull"  which would be covered in Collaborating section
- ...

**Reply by cc-a** [reply]
As briefly mentioned IRL, would be nice to have a consistent theme for this rather than just a smattering of fancier git stuff. Again based off of timings from the existing course the first two sections cover ~2 hrs material so not the most space for more content so I'm hesitant to add to what's already there. GitHub actions seems like a valuable addition that could be used in an overall collaboration theme and that also complements the software engineering course. Just that could easily take up an hour.

FWIW, I think multiple remotes is pretty rare. It's only ever come up a handful of times for me.

**Reply by AdrianDAlessandro** [reply]
I think perhaps a more complete / formal introduction to the GitHub collaborating workflow would add to what we already have. Perhaps with reference to [GitHub Flow](https://docs.github.com/en/get-started/quickstart/github-flow)? I also agree that CI is a very sensible addition

**Reply by dalonsoa** [reply]
I'd take the first to points mostly verbatim from the current course and then add:
- Stash (I find it invaluable when working on multiple stuff at the same time)
- Creating releases, which, as far as I know, can create tags along the way
- Continuous integration with GitHub Actions

I think the latter is really essential, but I find it difficult to marry with the overall theme of the examples of the course - recipes and book of recipes. There's not much to do with that in a CI system. Any ideas? Maybe running [`markdownlint`](https://github.com/marketplace/actions/markdownlint-mdl-action) in the book of recipes?

**Reply by dandavies99** [reply]
@dalonsoa Running `markdownlint` could be a nice way of showing CI in a course that is supposed to be fairly code agnostic.

Agree with @cc-a that a theme is important - perhaps "Intermediate Git and GitHub for Effective Collaboration"? Thinking about @AdrianDAlessandro's suggestion, it strikes me that many of the tools and commands we've suggested could be taught more directly in the context of collaborating with others, even though they are all useful generally. E.g.:

- Branches: Collaborators working on different features simultaneously.
- Forking, pull requests and git workflows: Similar to branches but working on someone else's code.
- Rebase and merge: Someone else has updated the main branch while you've been working.
- Reset: You've made a mistake and want to keep the commit history tidy for the benefit of collaborators.
- Stash: You want to checkout someone else's branch while you're working on your own.
- CI: Making sure every collaborator's contribution meets an agreed standard and doesn't break anything.
- Creating releases and adding tags: Communicating effectively to users/collaborators what has changed between versions of the code.

This is clearly too much content, so we could choose to signpost some of the topics without actually going through them. This could be partly determined by the exercise(s) that we incorporate...

**Reply by dalonsoa** [reply]
I think this list is a great summary of the intermediate course. I'm tempted to put it as it is in the business case, maybe leaving a couple of them more for them to follow up after a very brief introduction.

**Reply by dandavies99** [reply]
Sounds good. It would be nice to have "creating a release and adding tags" as a natural conclusion to this course? e.g. once you've made all the changes as a team, you summarise what you've done, ship it and celebrate.

**Reply by CWestICL** [reply]
```
stderr:
    C:\Users\cwest\.cache\pre-commit\repo2y8qakup\rbenv-default\bin\rbenv: line 44: cd: C:\Users\cwest\.cache\pre-commit\repo2y8qakup\rbenv-default\bin\rbenv: Not a directory
    rbenv: no such command `init'
```
Ran into an issue with markdownlint when running pre-commit. Might need investigation.

**Reply by alexdewar** [reply]
@CWestICL It could be something to do with your Ruby installation. Could try this: https://www.appsloveworld.com/ruby/100/49/rbenv-no-such-command-init

How did you install Ruby btw? Did you add anything to your ``PATH``?

**Reply by AdrianDAlessandro** [reply]
It's using pre-commit, which I believe should install all the necessary components by itself. It seems like it shouldn't need to have the rbenv on the path because it uses the full path to it anyway?

**Reply by alexdewar** [reply]
I seem to remember I had to install ruby explicitly when I did it but that was it; I guess ``pre-commit`` installs the needed ruby packages.

**Reply by AdrianDAlessandro** [reply]
We changed to using the Node.js version in that repo and it fixed the issue https://github.com/igorshubovych/markdownlint-cli#use-with-pre-commit

**Reply by alexdewar** [reply]
Oh that's cool. Which repo are you talking about btw?

Maybe we should start using the node.js version everywhere if it works out of the box (unlike the ruby version).

**Comment by dandavies99** [comment]
Hi everyone, I finally got around to making a start on splitting the course.

## Basic course

For the basic course, the [original repo](https://github.com/ImperialCollegeLondon/grad_school_git_course) can be used. There's a ➡️ [high level issue](https://github.com/ImperialCollegeLondon/grad_school_git_course/issues/34) ⬅️ with sub-issues that can be tackled individually. Feel free to add to the list of issues, claim them and start more detailed discussions within each one. 

## Intermediate course

For the intermediate course, there is a [new repo](https://github.com/ImperialCollegeLondon/intermediate_grad_school_git_course). Again, there's a ➡️ [high level issue](https://github.com/ImperialCollegeLondon/intermediate_grad_school_git_course/issues/1) ⬅️ but no sub-issues yet. I think we need to first decide on the high-level contents and structure. I've added [four lessons](https://github.com/ImperialCollegeLondon/intermediate_grad_school_git_course/tree/gh-pages/_episodes) based on the discussion above. I think these need a bit more discussion before we start working on them. Particularly, how we are going to incorporate / adapt the current collaboration exercise? What other exercises can we incorporate for the intermediate concepts and CI parts?
